### PR TITLE
(DOCSP-21141 & DOCSP-21137): Flutter SDK supported data types & create schema

### PR DESCRIPTION
**Pull request by mongodben**
## Pull Request Info

Docs for Flutter data types and modeling a Realm Object

### Jira

- https://jira.mongodb.org/browse/DOCSP-21141
- https://jira.mongodb.org/browse/DOCSP-21137

### Staged Changes (Requires MongoDB Corp SSO)

- [Data Types (Flutter)](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-21141/sdk/flutter/data-types/)
- [Define a Realm Object Schema (Flutter)](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-21141/sdk/flutter/define-realm-object-schema/)

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: examples/dart/test/data_types_test.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+part 'data_types_test.g.dart'; // :remove:
+// :uncomment-start:
+// part 'car.g.dart';
+// :uncomment-end:
+
+@RealmModel()
+class _Car {
+  @PrimaryKey()
+  late final int id; // Flutter SDK does not yet support ObjectId
```
Comment (blagoev) [inline examples/dart/test/data_types_test.dart:13]:
```suggestion
  late final int id;
```

Comment (blagoev) [reply examples/dart/test/data_types_test.dart:13]: ObjectId is just one type that can be a primary key. I think we should not steer people of using objectIds primary keys by default. In my experience developers start doing what the examples are using and always use the example code as the default. So its perfectly fine to have an int primary key here.

File: examples/dart/test/data_types_test.dart [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  final bool isElectric = false;
+  double milesDriven = 0;
+  late final List<String> attributes = [];
+  late _Person? owner;
+}
+
+@RealmModel()
+class _Person {
+  @PrimaryKey()
+  late final int id; // Flutter SDK does not yet support ObjectId
```
Comment (blagoev) [inline examples/dart/test/data_types_test.dart:25]:
```suggestion
  late final int id;
```

File: source/examples/generated/flutter/data_types_test.codeblock.data-types-example-model.dart [hunk 3]
```diff
@@ -0,0 +1,23 @@
+
+part 'car.g.dart';
+
+@RealmModel()
+class _Car {
+  @PrimaryKey()
+  late final int id; // Flutter SDK does not yet support ObjectId
```
Comment (blagoev) [inline source/examples/generated/flutter/data_types_test.codeblock.data-types-example-model.dart:7]:
```suggestion
  late final int id;
```

File: source/examples/generated/flutter/data_types_test.codeblock.data-types-example-model.dart [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  final bool isElectric = false;
+  double milesDriven = 0;
+  late final List<String> attributes = [];
+  late _Person? owner;
+}
+
+@RealmModel()
+class _Person {
+  @PrimaryKey()
+  late final int id; // Flutter SDK does not yet support ObjectId
```
Comment (blagoev) [inline source/examples/generated/flutter/data_types_test.codeblock.data-types-example-model.dart:19]:
```suggestion
  late final int id;
```

File: source/sdk/flutter/data-types.txt [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+Dart Types
+----------
+
+Realm supports the following Dart language data types:
+
+- ``int``
+- ``double``
+- ``bool``
+- ``String``
+- ``null``
```
Comment (blagoev) [inline source/sdk/flutter/data-types.txt:34]: I see we list `null` here. I suggest we remove it. It is not a type in the database, its more like a value for nullable properties in the data model.

File: source/sdk/flutter/data-types.txt [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+- ``List<T>``, where T is any of the supported data types
```
Comment (blagoev) [inline source/sdk/flutter/data-types.txt:35]:
```suggestion
- ``List<T>``, where T is any of the supported data types or a RealmObject type
```

File: source/sdk/flutter/data-types.txt [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+.. _flutter-embed-realm-object:
+
+Embed RealmObjects
```
Comment (blagoev) [inline source/sdk/flutter/data-types.txt:39]: We don't support embedded objects right now. If you mean we support object references we should reword this to be object references
so something like 
`You can also reference a RealmObject in another RealmObject`

File: examples/dart/test/data_types_test.dart [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+@RealmModel()
+class _Car {
+  @PrimaryKey()
+  late final int id; // Flutter SDK does not yet support ObjectId
+
+  String? licensePlate;
+  final bool isElectric = false;
+  double milesDriven = 0;
+  late final List<String> attributes = [];
```
Comment (blagoev) [inline examples/dart/test/data_types_test.dart:18]:
```suggestion
  late final List<String> attributes;
```

File: examples/dart/test/data_types_test.dart [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+// part 'car.g.dart';
+// :uncomment-end:
+
+@RealmModel()
+class _Car {
+  @PrimaryKey()
+  late final int id; // Flutter SDK does not yet support ObjectId
+
+  String? licensePlate;
+  final bool isElectric = false;
```
Comment (blagoev) [inline examples/dart/test/data_types_test.dart:16]:
```suggestion
  bool isElectric = false;
```

**Review by blagoev — CHANGES_REQUESTED** [review]
suggest some changes

**Review by blagoev — APPROVED** [review]

File: source/sdk/flutter/define-realm-object-schema.txt [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+      in the same package as the file you're currently working on.
+
+      .. literalinclude:: /examples/generated/flutter/schemas.codeblock.part-directive.dart
+         :language: dart
+         :caption: schemas.dart
+
+   .. step:: Create RealmModel
+
+      Create the model for your Realm schema in a private class. 
+      You must include the annotation `RealmModel <{+flutter-sdk-common+}/realm_common/RealmModel-class.html>`_
```
Comment (cbush) [inline source/sdk/flutter/define-realm-object-schema.txt:64]: Why not create a `:flutter-sdk-common:` role for these links? Using source constants in a URL is not a pattern I've seen before.

**Review by cbush — APPROVED** [review]

File: source/sdk/flutter/define-realm-object-schema.txt (hunk 10 above)
Comment (mongodben) [reply source/sdk/flutter/define-realm-object-schema.txt:64]: mainly b/c it's not a very common path to link to. as it's a different package from the main flutter package, which there is an inline role for `:flutter-sdk:`. 

i used the source constant instead following the pattern of the drivers team.

File: source/sdk/flutter/data-types.txt [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+.. contents:: On this page
+   :local:
+   :backlinks: none
+   :depth: 2
+   :class: singlecol
+
+The Flutter SDK currently only supports Dart-language data types. 
+The Flutter SDK **does not** support MongoDB.BSON types or Realm-specific types
+at the moment.
```
[the remark this answers is not in the document]
Comment (mongodben) [reply source/sdk/flutter/data-types.txt:17]: not currently scheduled in the sense that i couldn't find a ticket for it. 

but functionality exists for the more mature SDKs, so i assume it's coming to the Flutter SDK as well.

File: source/sdk/flutter/define-realm-object-schema.txt (hunk 10 above)
Comment (mongodben) [reply source/sdk/flutter/define-realm-object-schema.txt:64]: per slack convo, updating to be just links with `__`
